### ISSUE TITLE
Freeze<T> treats tuples as arrays

### ISSUE THREAD

**Issue by IchiganCS** [giusepperaso/structura.js#115]
First of all, thank you for this nice library!
Secondly, I think the default `Freeze<T>` type is not correctly implemented for tuples - I have created a small example and a potential (?) fix. From here, the default Freeze implementation is called `DefaultFreeze`, my new (incomplete) Freeze type is named Freeze

```ts
// just a few example types to test the implementation
type Test = [number, string]
type Test2 = number[]
type Test3 = {
    name: string,
    other: {
        name: number,
        ids: string[]
    }
}

// apply the freeze types to the examples
let x: Freeze<Test> = [3, 'test'] // correctly has type readonly [number, string]
let xDefault: DefaultFreeze<Test> = ['test', 3] // mistakenly has type readonly (string | number)[] and the code does (mistakenly) compile
let y: Freeze<Test2> = [3, 4] // correctly has type readonly number[]
let yDefault: DefaultFreeze<Test2> = [] // correctly has type readonly number[]
let z: Freeze<Test3> = { // also is correct for both Freeze and DefaultFreeze
    name: 'test',
    other: {
        name: 3,
        ids: ['id0', 'id1', 'id2']
    }
}

// with my Freeze implementation

type FreezeArr<T> = 
    T extends [infer First] ? readonly [Freeze<First>] : 
    T extends [infer First, infer Seconds] ? readonly [Freeze<First>, Freeze<Seconds>] : 
    T extends [infer First, ...infer Arr] ?  readonly [Freeze<First>, FreezeArr<[...Arr]>] : 
    T extends (infer First)[] ? readonly (Freeze<First>)[] :
    never
type FreezeObj<T extends object> = {
    readonly [Name in keyof T]: Freeze<T[Name]>
}
type Freeze<T> = T extends [...infer Tup] ? FreezeArr<[...Tup]> : T extends object ? FreezeObj<T> : T
```

I hope this library can implement this behavior in some way, because currently I cannot use the library for my project (although I would like to), because there are many tupled types in my project.

Best regards!

**Comment by IchiganCS** [comment]
I have added a pull request to fix this issue.

**Comment by giusepperaso** [comment]
Sorry if I was very slow
The fix is very good, thanks a lot!

(it is available from version 0.12.6)
